### PR TITLE
UCP: tm offload rndv support

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe pls take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1753/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3691/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1757/ for details.

File: src/ucp/tag/offload.c [hunk 1]
```diff
@@ -44,6 +44,66 @@ void ucp_tag_offload_completed(uct_tag_context_t *self, uct_tag_t stag,
...
+    ucp_rndv_matched(req->recv.worker, req, &rts);
+}
+
+ucs_status_t ucp_tag_offload_unexp_rndv(void *arg, unsigned flags,
+                                        uint64_t stag, const void *hdr,
+                                        unsigned hdr_length,
+                                        uint64_t remote_addr, size_t length,
+                                        const void *rkey_buf)
+{
+    ucp_rndv_rts_hdr_t *rts = (ucp_rndv_rts_hdr_t*)(((ucp_tag_hdr_t*)hdr) - 1);
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: isn't this illegal to access memory before 'hdr'?

File: src/ucp/tag/offload.c [hunk 2]
```diff
@@ -213,6 +274,81 @@ static ucs_status_t ucp_tag_offload_eager_zcopy(uct_pending_req_t *self)
...
+        .length            = req->send.length
+    };
+
+    return uct_ep_tag_rndv_request(ep->uct_eps[req->send.lane], req->send.tag,
+                                   &rndv_hdr, sizeof(rndv_hdr));
+}
+
+ucs_status_t ucp_tag_offload_rndv_zcopy(uct_pending_req_t *self)
+{
+    void *rndv_op;
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: pls move this variable to be last, so variables with initializers are first

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1758/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3695/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1759/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3696/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1767/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3705/ for details (Mellanox internal link).

File: src/ucp/tag/offload.c [hunk 3]
```diff
@@ -44,21 +44,88 @@ void ucp_tag_offload_completed(uct_tag_context_t *self, uct_tag_t stag,
...
-
-    ucp_iface = ucs_queue_head_elem_non_empty(&ctx->tm.offload_ifaces,
-                                              ucp_worker_iface_t, queue);
-    ucp_request_memory_dereg(ctx, ucp_iface->rsc_index, req->recv.datatype,
-                             &req->recv.state);
-    status = uct_iface_tag_recv_cancel(ucp_iface->iface, &req->recv.uct_ctx, force);
-    if (status != UCS_OK) {
-        ucs_error("Failed to cancel recv in the transport: %s",
-                  ucs_status_string(status));
+    if (req->flags & UCP_REQUEST_FLAG_OFFLOADED) {
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: if (!offloaded) return

File: src/ucp/tag/offload.c [hunk 4]
```diff
@@ -44,21 +44,88 @@ void ucp_tag_offload_completed(uct_tag_context_t *self, uct_tag_t stag,
...
+                             ucs_status_t status)
+{
+    ucp_request_t *req        = ucs_container_of(self, ucp_request_t, recv.uct_ctx);
+    ucp_context_t *ctx        = req->recv.worker->context;
+    ucp_sw_rndv_hdr_t *sreq   = (ucp_sw_rndv_hdr_t*)header;
+    ucp_worker_iface_t *iface = ucs_queue_head_elem_non_empty(&ctx->tm.offload_ifaces,
+                                                              ucp_worker_iface_t, queue);
+    ucp_rndv_rts_hdr_t rts;
+
+    rts.sreq      = sreq->super;
```
Comment (yosefe) [inline src/ucp/tag/offload.c]: pls add comment like "emulate rts without key"

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1784/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3721/ for details (Mellanox internal link).

**Pull request merged**